src/libraries/System.Text.RegularExpressions/gen/RegexAnalyzer.cs

```diff
@@ -0,0 +1,416 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace System.Text.RegularExpressions.Generator
+{
+    /// <summary>
+    /// Roslyn analyzer that searches for invocations of the Regex constructors, or the
+    /// Regex static methods and analyzes if the callsite could be using the Regex Generator instead.
+    /// If so, it will emit an informational diagnostic to suggest use the Regex Generator.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class RegexAnalyzer : DiagnosticAnalyzer
+    {
+        // private members
+        private const string RegexTypeName = "System.Text.RegularExpressions.Regex";
+        private const string RegexGeneratorTypeName = "System.Text.RegularExpressions.RegexGeneratorAttribute";
+        private const string TimeSpanTypeName = "System.TimeSpan";
+        private const string TimeoutTypeName = "System.Threading.Timeout";
+
+        // internal members
+        internal const string PatternIndexName = "PatternIndex";
+        internal const string RegexOptionsIndexName = "RegexOptionsIndex";
+        internal const string RegexTimeoutIndexName = "RegexTimeoutIndex";
+        internal const string RegexTimeoutName = "RegexTimeout";
+        internal const string DiagnosticId = "SYSLIB1046";
+
+        /// <inheritdoc />
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(DiagnosticDescriptors.UseRegexSourceGeneration);
+
+        /// <inheritdoc />
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+
+            // Register analysis of calls to the Regex constructors
+            context.RegisterOperationAction(AnalyzeObjectCreation, OperationKind.ObjectCreation);
```

Comment:
Typically these registrations aren't done directly in Initialize, but instead are done inside of another registration, e.g. context.RegisterCompilationStartAction, and then in that lambda prior to doing these RegisterOperationAction calls, you'd do the checking to make sure all the relevant support is available (i.e. the work you do in https://github.com/dotnet/runtime/pull/68976/files#diff-e0f48f271338a3d805aa3fce6ce76d65231da7df898f5b27b7ae74e1cea85b95R393 to look up types in the compilation should be done here instead, once rather than per operation), such that we don't pay any of the per-operation costs for this analyzer if the relevant support isn't available anyway.

Reply:
makes sense, I can do that.